docs/source/extending/extending_models.rst

```diff
@@ -0,0 +1,240 @@
+    from pykeen.triples import TriplesFactory
+
+    class ModifiedLinearDistMult(EntityRelationEmbeddingModel):
+        loss_default = NSSALoss
+
+        def __init__(
+            self,
+            hidden_dim: int = 20,  # extra stuff!
+            **kwargs,  # pass everything else, you neither have to understand nor be able to handle the truth
+        ):
+            super().__init__(**kwargs)
+
+            # Save some extra state information
+            self.hidden_dim = hidden_dim
+
+            # Note that the ``embedding_dim`` is available to all EntityRelationEmbeddingModels after init.
+            self.linear1 = torch.nn.Linear(self.embedding_dim, self.hidden_dim)
+            self.linear2 = torch.nn.Linear(self.hidden_dim, self.embedding_dim)
+
+        def score_hrt(self, hrt_batch):
+            h = self.entity_embeddings(hrt_batch[:, 0])
+            r = self.relation_embeddings(hrt_batch[:, 1])
+            t = self.entity_embeddings(hrt_batch[:, 2])
+
+            # add more transformations
+            h = self.linear2(self.linear1(h))
+            t = self.linear2(self.linear1(t))
+
+            return h * r.sigmoid() * t
+
+        def _reset_parameters_(self):  # noqa: D102
+            super()._reset_parameters_()
+
+            # weight initialization
+            torch.nn.init.zeros_(self.linear1.bias)
+            torch.nn.init.zeros_(self.linear2.bias)
+            torch.nn.init.xavier_uniform_(self.linear1.weight)
+            torch.nn.init.xavier_uniform_(self.linear2.weight)
+
+Adding Custom HPO Default Ranges
+--------------------------------
+All subclasses of :class:`pykeen.models.base.Model` can specify the default
+ranges or values used during hyper-parameter optimization (HPO). PyKEEN
+implements a simple dictionary-based configuration that is interpreted
+by :func:`pykeen.hpo.hpo.suggest_kwargs` in the HPO pipeline.
+
+HPO default ranges can be applied to all keyword arguments appearing in the
+``__init__()`` function of your model by setting a class-level variable called
+``hpo_default``.
+
+For example, the ``hidden_dim`` can be specified as being on a range between
+15 and 50 with the following:
+
+.. code-block:: python
+
+    class ModifiedLinearDistMult(EntityRelationEmbeddingModel):
+        hpo_default = {
+            'hidden_dim': dict(type=int, low=15, high=50)
+        }
+        ...
+
+A step size can be imposed with ``q``:
+
+.. code-block:: python
+
+    class ModifiedLinearDistMult(EntityRelationEmbeddingModel):
+        hpo_default = {
+            'hidden_dim': dict(type=int, low=15, high=50, q=5)
+        }
+        ...
+
+An alternative scale can be imposed with ``scale``. Right now, the
+default is linear, and ``scale`` can optionally be set to ``power_two``
+for integers as in:
+
+.. code-block:: python
+
+    class ModifiedLinearDistMult(EntityRelationEmbeddingModel):
+        hpo_default = {
+            'hidden_dim': dict(type=int, low=15, high=50, scale='power_two')
```

Comment:
Maybe it is better to adapt the bounds in the example since values would be sampled from the interval 2^15-2^50.

Reply:
Fixed in a6afff03